logs/util/read_logs_blocker.go

```diff
@@ -0,0 +1,60 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+package util
+
+import (
+	"sync"
+)
+
+var logBlockerSingleton *LogBlocker
+
+type LogBlocker struct {
+	maxLogBufferSize int64
+	logsBufferSize int64
+	logsBufferSizeMutex sync.Mutex
+}
+
+func GetLogBlocker() *LogBlocker {
+	if logBlockerSingleton == nil {
+		logBlockerSingleton = &LogBlocker{
+			maxLogBufferSize: int64(-1),
+			logsBufferSize: 0,
+			logsBufferSizeMutex: sync.Mutex{},
+		}
+	}
+	return logBlockerSingleton
+}
+
+func (l *LogBlocker) Add(v int64) {
+	l.logsBufferSizeMutex.Lock()
+	defer l.logsBufferSizeMutex.Unlock()
+	l.logsBufferSize = l.logsBufferSize + v
+}
+
+func (l *LogBlocker) Subtract(v int64) {
+	l.logsBufferSizeMutex.Lock()
+	defer l.logsBufferSizeMutex.Unlock()
+	l.logsBufferSize = l.logsBufferSize - v
+}
+
+// Reset this is test only code
+// do not use for application
+func (l *LogBlocker) Reset() {
+	l.logsBufferSizeMutex.Lock()
+	defer l.logsBufferSizeMutex.Unlock()
+	l.logsBufferSize = 0
+}
+
+func (l *LogBlocker) SetMaxLogBuffer(v int64) {
+	l.maxLogBufferSize = v
+}
+
+func (l *LogBlocker) Block() (block bool, bufferSize int64, maxBufferSize int64) {
```

Comment:
It is not obvious to me what a LogBlocker does.

Reply:
I doubt that will work because the weighted semaphore will try to acquire lock on addition. We need to force acquire lock because we add the header size at and change the state size for truncating. Thus, we need to have a flexible size that allows the exporting path to add size when it needs to. So, this is not really a hard stop at 1mb (configurable) it is a soft spot that blocks the next execution of reading and opening new files. Since the exporter should constantly change line size state.

Reply:
Unfortunately, it is up to the reader to stop itself. Instead of fully blocking.

Reply:
A log event can already be added to the buffer, but before flushing, the size will change.